manifests/server.pp

```diff
@@ -34,28 +38,73 @@
 # @author InnoGames GmbH
 #
 class clickhouse::server (
-    String[1]        $package_name   = 'clickhouse-server',
-    String[1]        $package_ensure = 'installed',
-    String[1]        $service_name   = $package_name,
+    String[1]                  $package_name   = 'clickhouse-server',
+    String[1]                  $package_ensure = 'installed',
+    String[1]                  $service_name   = $package_name,
     Variant[Boolean, Enum[
         'running',
         'stopped'
-    ]]               $service_ensure = 'running',
+    ]]                         $service_ensure = 'running',
     Variant[Boolean, Enum[
         'manual',
         'mask'
-    ]]               $service_enable = true,
-    Stdlib::Unixpath $conf_dir       = '/etc/clickhouse-server',
-    Stdlib::Unixpath $conf_d_dir     = "${conf_dir}/conf.d",
-    Stdlib::Unixpath $users_d_dir    = "${conf_dir}/users.d",
+    ]]                         $service_enable = true,
+    Optional[Stdlib::Unixpath] $conf_dir       = undef,
+    Stdlib::Unixpath           $config_dir     = $conf_dir ? {
+        undef   => '/etc/clickhouse-server',
+        default => $conf_d_dir,
+    },
+    Optional[Stdlib::Unixpath] $conf_d_dir     = undef,
+    Stdlib::Unixpath           $config_d_dir   = $conf_d_dir ? {
+        undef   => "${config_dir}/config.d",
+        default => $conf_d_dir,
+    },
+    Stdlib::Unixpath           $users_d_dir    = "${config_dir}/users.d",
 ) inherits clickhouse {
 
+    # TODO: remove it in 3 releases
+    if ($conf_d_dir) {
+        notify { 'conf_d deprecation':
+            message  => @("END")
+                Parameter conf_d_dir is deprecated, use config_d_dir
+                Be aware that after remove it the ${conf_d_dir} will be removed as well, you shouldn't use it.
+                See https://clickhouse.yandex/docs/en/operations/configuration_files
+                | END
+            ,
+            loglevel => 'warning',
+        }
+
+        file { $conf_d_dir:
+            ensure  => 'directory',
+            recurse => true,
+            purge   => true,
```

Comment:
No, there could be manually managed configs if a user starts to use this module. We shouldn't delete them blindly.

Reply:
how? `purge` is already true

Reply:
Without `force`, `purge` + `recurse` doesn't delete directory with existing files. Here's example:

$ pwd
/etc/clickhouse-server
$ sudo mkdir conf.d -p
$ sudo touch conf.d/config-to-be-removed
$ sudo puppetrun
.....
Notice: /Stage[main]/Clickhouse::Server/File[/etc/clickhouse-server/conf.d]: Not removing directory; use 'force' to override
Notice: /Stage[main]/Clickhouse::Server/File[/etc/clickhouse-server/conf.d]/ensure: removed (corrective)
.....
$ ls -l conf.d/
total 0
-rw-r--r-- 1 root root 0 Jul 22 08:50 config-to-be-removed
```

That's why it is in the current state. Users should puppetize existing files before lose them.

Reply:
Ok, I somehow mixed-up the things in my head. Not sure about `force` still, since it could remove directories. They should not affect the server configuration but could be there for reasons.

I guess you are right, and they make sense. I'll add `force` here. And let's discuss conf.d in another branch.

Reply:
added